### PR TITLE
[FLINK-11480][hive] Create HiveTableFactory that creates TableSource/…

### PR DESCRIPTION
**Pull request by xuefuz**
…Sink from a Hive table

## What is the purpose of the change

Create HiveTableFactory that creates TableSource/Sink from a Hive table

## Brief change log

*(for example:)*
  - Created a new class called HiveTableFactory for tables stored in Hive metastore
  - Add a new  catalog API getTableFactory() that optionally returns a TableFactory instance

## Verifying this change

Added a test that's going to be enabled when a dependent problem is fixed.

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (avaDocs)

### CODE REVIEW

File: flink-connectors/flink-connector-hive/pom.xml [hunk 1]
```diff
@@ -59,6 +59,12 @@ under the License.
 			<version>${project.version}</version>
 			<scope>provided</scope>
 		</dependency>
+                <dependency>
+                        <groupId>org.apache.flink</groupId>
+                        <artifactId>flink-table-api-java-bridge_2.11</artifactId>
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/pom.xml]:
```suggestion
                        <artifactId>flink-table-api-java-bridge_${scala.binary.version}</artifactId>
```

BTW, since we depend on flink-table-api-java-bridge now, we can probably just remove the above dependency of flink-table-api-java

File: flink-connectors/flink-connector-hive/pom.xml [hunk 2]
```diff
@@ -59,6 +59,12 @@ under the License.
 			<version>${project.version}</version>
 			<scope>provided</scope>
 		</dependency>
+                <dependency>
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/pom.xml]: the indentation looks not right. Not sure if it's a problem of github. Can you double check?

File: flink-connectors/flink-connector-hive/pom.xml [hunk 3]
```diff
@@ -59,6 +59,12 @@ under the License.
...
+                        <groupId>org.apache.flink</groupId>
+                        <artifactId>flink-table-api-java-bridge_2.11</artifactId>
+                        <version>${project.version}</version>
+                        <scope>compile</scope>
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/pom.xml]: we can just remove the scope tag since the default scope is compile.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableFactory.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+import org.apache.flink.table.sources.InputFormatTableSource;
+import org.apache.flink.table.sources.TableSource;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Preconditions;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A table factory implementation\ for tables stored in Hive catalog.
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableFactory.java]:
```suggestion
 * A table factory implementation for tables stored in Hive catalog.
```

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/batch/connectors/hive/HiveTableFactoryTest.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+		catalog = HiveTestUtils.createHiveCatalog();
+		catalog.open();
+	}
+
+	@AfterClass
+	public static void close() throws IOException {
+		catalog.close();
+	}
+
+	// Enable this test after FLINK-12892 is fixed.
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/batch/connectors/hive/HiveTableFactoryTest.java]: FYI, FLINK-12892 is merged

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/Catalog.java [hunk 6]
```diff
@@ -43,6 +45,10 @@
 @PublicEvolving
 public interface Catalog {
```
Comment (bowenli86) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/Catalog.java]: minor: add javadoc?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableSinkFactory.java [hunk 7]
```diff
@@ -39,4 +40,9 @@
 	 * @return the configured table sink.
 	 */
 	TableSink<T> createTableSink(Map<String, String> properties);
+
+	default TableSink<T> createTableSink(CatalogTable table) {
```
Comment (bowenli86) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableSinkFactory.java]: add javadoc?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableSourceFactory.java [hunk 8]
```diff
@@ -39,4 +40,9 @@
 	 * @return the configured table source.
 	 */
 	TableSource<T> createTableSource(Map<String, String> properties);
+
+	default TableSource<T> createTableSource(CatalogTable table) {
```
Comment (bowenli86) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableSourceFactory.java]: add javadoc?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/catalog/DatabaseCalciteSchema.java [hunk 9]
```diff
@@ -77,6 +84,23 @@ public Table getTable(String tableName) {
 						!connectorTable.isBatch(),
 						FlinkStatistic.of(tableSource.getTableStats().orElse(null))))
 					.orElseThrow(() -> new TableException("Cannot query a sink only table."));
+			} else if (table instanceof CatalogTable) {
+				Optional<TableFactory> tableFactory = catalog.getTableFactory();
+				TableSource<Row> tableSource = tableFactory.map(tf -> ((TableSourceFactory) tf).createTableSource((CatalogTable) table))
```
Comment (bowenli86) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/catalog/DatabaseCalciteSchema.java]: the `tf` in `.map(tf -> xxx)` is an Optional object and cannot be casted to `TableSourceFactory`, right? 

BTW, what happens if the `tableFactory` is not present? Also, maybe I missed it - I only find table source, what about table sink?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/catalog/DatabaseCalciteSchema.java [hunk 10]
```diff
@@ -77,6 +84,23 @@ public Table getTable(String tableName) {
...
+					.orElse(TableFactoryUtil.findAndCreateTableSource(((CatalogTable) table).toProperties()));
+
+				if (!(tableSource instanceof StreamTableSource)) {
```
Comment (bowenli86) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/catalog/DatabaseCalciteSchema.java]: the message mentions `BoundedTableSource`, so shall we check for it?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/catalog/DatabaseCalciteSchema.java (hunk 9 above)
Comment (xuefuz) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/catalog/DatabaseCalciteSchema.java]: map() is applied to an Optional object, so the key (tf) is a TableFactory instance.
As you can see from other types of tables, calcite seems only need TableSource.
If tableFactory isn't present, then orElse() clause kicks in. The test verifies that.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/catalog/DatabaseCalciteSchema.java (hunk 10 above)
Comment (xuefuz) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/catalog/DatabaseCalciteSchema.java]: The message was from Dawid's change. I didn't add it.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/catalog/DatabaseCalciteSchema.java (hunk 9 above)
Comment (bowenli86) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/catalog/DatabaseCalciteSchema.java]: nice, I didn't know map() can be applied to Optional objects

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java [hunk 11]
```diff
@@ -169,6 +172,11 @@ public void close() throws CatalogException {
 		}
 	}
 
+	@Override
+	public Optional<TableFactory> getTableFactory() {
+		return Optional.of(new HiveTableFactory());
```
Comment (lirui-apache) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java]: Does this mean `HiveTableFactory ` will create source/sink for both generic and hive tables?

Comment (xuefuz) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java]: Yes. However, for generic tables, it delegates the responsibility to auto tablefactory discovery service, as you can see from HiveTableFactory implementation.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/catalog/DatabaseCalciteSchema.java (hunk 10 above)
Comment (xuefuz) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/catalog/DatabaseCalciteSchema.java]: Actually the message has changed. BoundedTableSource isn't referred any more.

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/batch/connectors/hive/HiveTableFactoryTest.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/batch/connectors/hive/HiveTableFactoryTest.java]: remove this comment?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/AbstractCatalogTable.java [hunk 13]
```diff
@@ -94,10 +93,11 @@ public String getComment() {
 		Map<String, String> properties = getProperties();
 		properties.remove(GenericInMemoryCatalog.FLINK_IS_GENERIC_KEY);
 
-		descriptor.putPropertiesWithPrefix(CatalogTableConfig.TABLE_PROPERTIES, properties);
+//		descriptor.putPropertiesWithPrefix(CatalogTableConfig.TABLE_PROPERTIES, properties);
+		descriptor.putProperties(properties);
 
-		descriptor.putString(CatalogTableConfig.TABLE_COMMENT, getComment());
-		descriptor.putString(CatalogTableConfig.TABLE_PARTITION_KEYS, String.join(",", partitionKeys));
+//		descriptor.putString(CatalogTableConfig.TABLE_COMMENT, getComment());
```
Comment (bowenli86) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/AbstractCatalogTable.java]: shall we just remove them completely?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/TableFactoryUtil.java [hunk 14]
```diff
@@ -47,15 +48,14 @@ public static ExternalCatalog findAndCreateExternalCatalog(Descriptor descriptor
 	 */
 	public static <T> TableSource<T> findAndCreateTableSource(Descriptor descriptor) {
 		Map<String, String> properties = descriptor.toProperties();
-
```
Comment (bowenli86) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/TableFactoryUtil.java]: nit: revert this unnecessary change?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/TableFactoryUtil.java [hunk 15]
```diff
@@ -83,4 +86,19 @@ public static ExternalCatalog findAndCreateExternalCatalog(Descriptor descriptor
 
 		return tableSink;
 	}
+
+	/**
+	 * Returns a table sink matching the descriptor.
+	 */
+	public static <T> TableSink<T> findAndCreateTableSink(CatalogTable table) {
```
Comment (bowenli86) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/TableFactoryUtil.java]: I actually wonder if we can just call `findAndCreateTableSink(table.toProperties()) ` in `HiveTableFactory` rather than adding the two new utils? `TableFactoryUtil` just need a set of properties, and doesn't seem to need to understand how a CatalogTable serialize itself into a set of properties

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/Catalog.java [hunk 16]
```diff
@@ -43,6 +45,16 @@
...
 
+	/**
+	 * Get an optional {@link TableFactory} instance that's responsible to generate source/sink for tables
```
Comment (bowenli86) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/Catalog.java]:
```suggestion
	 * Get an optional {@link TableFactory} instance that's responsible for generating source/sink for tables
```

Also, can you mention in javadoc when and where this is supposed to be called? Audience may get confused on how this cooperates with table discovery service.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/AbstractCatalogTable.java [hunk 17]
```diff
@@ -94,10 +93,11 @@ public String getComment() {
 		Map<String, String> properties = getProperties();
 		properties.remove(GenericInMemoryCatalog.FLINK_IS_GENERIC_KEY);
 
-		descriptor.putPropertiesWithPrefix(CatalogTableConfig.TABLE_PROPERTIES, properties);
+//		descriptor.putPropertiesWithPrefix(CatalogTableConfig.TABLE_PROPERTIES, properties);
```
Comment (bowenli86) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/AbstractCatalogTable.java]: just remove this line?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableFactory.java [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+import java.util.Map;
+
+/**
+ * A table factory implementation for tables stored in Hive catalog.
+ */
+public class HiveTableFactory implements TableSourceFactory<Row>, TableSinkFactory<Row> {
+
+	@Override
+	public Map<String, String> requiredContext() {
+		return null;
```
[the remark this answers is not in the document]
Comment (xuefuz) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableFactory.java]: No, HiveTableFactory isn't participating in table factory discovery service. It's returned by HiveCatalog and used for tables stored there.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/TableFactoryUtil.java (hunk 15 above)
Comment (xuefuz) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/TableFactoryUtil.java]: Well, there is just one option. In that case, we need to make the other two util methods public, which doesn't see good. I'd think the way here is cleaner.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/TableFactoryUtil.java (hunk 14 above)
Comment (xuefuz) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/TableFactoryUtil.java]: I guess it's okay. It's more consistent with findAndCreateTableSink().

File: flink-connectors/flink-connector-hive/pom.xml [hunk 19]
```diff
@@ -41,18 +41,26 @@ under the License.
 
 		<dependency>
 			<groupId>org.apache.flink</groupId>
-			<artifactId>flink-table-api-java-bridge_${scala.binary.version}</artifactId>
+			<artifactId>flink-table-common</artifactId>
 			<version>${project.version}</version>
 			<scope>provided</scope>
 		</dependency>
 
+
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/pom.xml]: revert extra line

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableFactory.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (bowenli86) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableFactory.java]: right, so shall we throw UnsupportedOperationException in `requiredContext()` and `supportedProperties()`?

File: flink-connectors/flink-connector-hive/pom.xml (hunk 19 above)
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/pom.xml]: revert this extra line?

**Comment by bowenli86** [comment]
there are some failures in flink-python's table API tests and flink-table-api-java. Can you fix them?

**Comment by bowenli86** [comment]
LGTM, will merge once the build passes

**Comment by bowenli86** [comment]
```
01:04:18.649 [ERROR] src/test/java/org/apache/flink/table/catalog/AbstractCatalogTableTest.java:[23,8] (imports) UnusedImports: Unused import: org.apache.flink.table.catalog.config.CatalogTableConfig.
```

@xuefuz seems there's a compiling error, can you fix it?

**Comment by bowenli86** [comment]
FYI, I just realized that we probably need to update the factory interfaces to take a `CatalogBaseTable` rather than `CatalogTable` to include views. 

That can be done in a following PR.

**Comment by xuefuz** [comment]
Actually I don't think TableFactory takes care of views. It only generate source/sink for real tables. Views will be expanded first and sources be generated only for real tables referenced in the view. This part is not the responsibility of TableFactory.
